notebooks/02-Distributed-Training_comments.ipynb

```diff
@@ -0,0 +1,962 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import shutil"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "model_registry = \"/mnt/user_storage/mlflow/doggos\"\n",
+    "os.path.isdir(model_registry) and shutil.rmtree(model_registry)  # clean up\n",
+    "os.makedirs(model_registry, exist_ok=True)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Training"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We'll define our training workload by specifying our:\n",
+    "- experiment and model parameters\n",
+    "- compute scaling configuration\n",
+    "- forward pass for batches of training and validation data\n",
+    "- train loop for each epoch of data (and checkpointing)\n",
+    "\n",
+    "<img src=\"https://raw.githubusercontent.com/anyscale/foundational-ray-app/refs/heads/main/images/trainer.png\" width=500>"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Train loop config\n",
+    "experiment_name = \"doggos\"\n",
+    "train_loop_config = {\n",
+    "    \"model_registry\": model_registry,\n",
+    "    \"experiment_name\": experiment_name,\n",
+    "    \"embedding_dim\": 512,\n",
+    "    \"hidden_dim\": 256,\n",
+    "    \"dropout_p\": 0.3,\n",
+    "    \"lr\": 1e-3,\n",
+    "    \"lr_factor\": 0.8,\n",
+    "    \"lr_patience\": 3,\n",
+    "    \"num_epochs\": 20,\n",
+    "    \"batch_size\": 256,\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Scaling config\n",
+    "num_workers = 4\n",
+    "scaling_config = ray.train.ScalingConfig(\n",
+    "    num_workers=num_workers,\n",
+    "    use_gpu=True,\n",
+    "    resources_per_worker={\"CPU\": 8, \"GPU\": 1})"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import tempfile\n",
+    "import mlflow\n",
+    "import numpy as np\n",
+    "from ray.train.torch import TorchTrainer"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def train_epoch(ds, batch_size, model, num_classes, loss_fn, optimizer):\n",
+    "    model.train()\n",
+    "    loss = 0.0\n",
+    "    ds_generator = ds.iter_torch_batches(batch_size=batch_size, collate_fn=collate_fn)\n",
+    "    for i, batch in enumerate(ds_generator):\n",
+    "        optimizer.zero_grad()  # reset gradients\n",
+    "        z = model(batch)  # forward pass\n",
+    "        targets = F.one_hot(batch[\"label\"], num_classes=num_classes).float()\n",
```

Comment:
nit: I think cross entropy can take in just the batch["label"] index directly. No need for the explicit onehot conversion